src/plugins/kibana/public/settings/sections/indices/add_data_steps/pipeline_setup/lib/pipeline.js

```diff
@@ -0,0 +1,145 @@
+import _ from 'lodash';
+import * as ProcessorTypes from './processor_types';
+
+export default class Pipeline {
+
+  constructor() {
+    this.processors = [];
+    this.counter = 0;
+    this.input = {};
+    this.output = undefined;
+    this.dirty = false;
+
+    this._processorTypes = _.reduce(ProcessorTypes, (result, Type, key) => {
+      const instance = new Type();
+      result[instance.typeId] = Type;
+      return result;
+    }, {});
+  }
+
+  get model() {
+    return {
+      input: this.input,
+      processors: _.map(this.processors, processor => processor.model)
+    };
+  }
+
+  load(pipeline) {
+    while (this.processors.length > 0) {
+      this.processors.pop();
+    }
+
+    pipeline.processors.forEach((processor) => {
+      this.addExisting(processor);
+    });
+  }
+
+  remove(processor) {
+    const processors = this.processors;
+    const index = processors.indexOf(processor);
+
+    processors.splice(index, 1);
+  }
+
+  moveUp(processor) {
+    const processors = this.processors;
+    const index = processors.indexOf(processor);
+
+    if (index === 0) return;
+
+    const temp = processors[index - 1];
+    processors[index - 1] = processors[index];
+    processors[index] = temp;
+  }
+
+  moveDown(processor) {
+    const processors = this.processors;
+    const index = processors.indexOf(processor);
+
+    if (index === processors.length - 1) return;
+
+    const temp = processors[index + 1];
+    processors[index + 1] = processors[index];
+    processors[index] = temp;
+  }
+
+  addExisting(existingProcessor) {
+    const Type = this._processorTypes[existingProcessor.typeId];
```

Comment:
Did `const Type = existingProcessor.constructor` not work for some reason? I would have expected it to, and if it did this would be much simpler and eliminate the need for this _processorTypes property.